Java/serialization/src/main/java/com/amazon/randomcutforest/serialize/json/v1/V1JsonToV2StateConverter.java

```diff
@@ -154,7 +135,69 @@ public void addSampler(V1SerializedRandomCutForest.Sampler sampler) {
             samplerState.setMaxSequenceIndex(sampler.getEntriesSeen());
             samplerState.setInitialAcceptFraction(1.0);
 
-            compactSamplerStates.add(samplerState);
+            if (compactSamplerStates.size() < maxNumberOfTrees) {
+                compactSamplerStates.add(samplerState);
+            }
         }
     }
+
+    /**
+     * the function merges a collection of RCF-1.0 models with same model parameters
+     * and fixes the number of trees in the new model (which has to be less or equal
+     * than the sum of the old models) The conversion uses the execution context of
+     * the first forest and can be adjusted subsequently by setters
+     * 
+     * @param serializedForests A non-empty list of forests (together having more
+     *                          trees than numberOfTrees)
+     * @param numberOfTrees     the new number of trees
+     * @param precision         the precision of the new forest
+     * @return a merged RCF with the first numberOfTrees trees
+     */
+    public RandomCutForestState convert(List<V1SerializedRandomCutForest> serializedForests, int numberOfTrees,
+            Precision precision) {
+        checkArgument(serializedForests.size() > 0, "incorrect usage of convert");
+        checkArgument(numberOfTrees > 0, "incorrect parameter");
+        int sum = 0;
+        for (int i = 0; i < serializedForests.size(); i++) {
+            sum += serializedForests.get(i).getNumberOfTrees();
+        }
+        checkArgument(sum >= numberOfTrees, "incorrect parameters");
```

Comment:
To answer your first question : Can you stop adding trees when it is larger than or equals numberOfTrees? This code will stop at 30 trees -- take a look at the tests :)

Reply:
The coordinator should collect all the different model partitions, decide on how many trees it wants (has to be less than the sum of the trees across the different partitions) and call this. If the number of trees in the existing models do not meet the default threshold then the number of trees should be set to that lower number. Basically the coordinator should be the single repository of "the state of the merge". Otherwise there will be all sorts of issues with consistency.

Reply:
The caller won’t know the actual number of trees in the json before invoking your method.  I want to avoid the illegalargument exception as exceptions is more expensive than an if else check.

Do you think this is better? I would return earlier if I know the conversion won’t succeed.  Change the number of trees to min(), but it brings complications that we need a threshold.  For example, if the total number of trees in the checkpoint is 3, we may not want to continue.

```
public Optional<RandomCutForestState> convert(ArrayList<String> jsons, int numberOfTrees, Precision precision)
            throws IOException {
        ArrayList<V1SerializedRandomCutForest> forests = new ArrayList<>();
    int sum = 0;
        for (int i = 0; i < jsons.size(); i++) {
       V1SerializedRandomCutForest forest = mapper.readValue(jsons.get(i), V1SerializedRandomCutForest.class)
            forests.add(forest);
      sum += forest.getNumberOfTrees();
        }
    if (sum < numberOfTrees) {
      return Optional.empty();
    }
        return Optional.ofNullable(convert(forests, numberOfTrees, precision));
    }
```

Reply:
about the coordinator should be the single repository: we discussed offline.  Our conclusion so far is: in single stream AD, coordinating node will be the single threaded owner.  The model node will send different messages depending on the coordinating is old or new.  if the number of trees is different, it will send ExceptionA to the new coordinating node; but send scores to the old node.  The new coordinating node will get ExceptionA and try to merge or cold start. In HCAD, the worker node itself will do the conversion as each entity model is located in one node.

Reply:
Done